iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/digitaltwin/DigitalTwinClientTests.java

```diff
@@ -225,6 +235,56 @@ public void getDigitalTwinWithAzureSasCredential() {
         assertEquals(responseWithHeaders.body().getMetadata().getModelId(), E2ETestConstants.THERMOSTAT_MODEL_ID);
     }
 
+    @Test
+    public void digitalTwinClientTokenRenewalWithAzureSasCredential()
+    {
+        if (protocol != MQTT)
+        {
+            // This test is for the service client, so no need to rerun it for all the different device protocols
+            return;
+        }
+
+        IotHubConnectionString iotHubConnectionStringObj =
+            IotHubConnectionStringBuilder.createIotHubConnectionString(IOTHUB_CONNECTION_STRING);
+
+        IotHubServiceSasToken serviceSasToken = new IotHubServiceSasToken(iotHubConnectionStringObj);
+        AzureSasCredential sasCredential = new AzureSasCredential(serviceSasToken.toString());
+        digitalTwinClient = new DigitalTwinClient(iotHubConnectionStringObj.getHostName(), sasCredential);
+
+        // get a digital twin with a valid SAS token in the AzureSasCredential instance
+        // don't care about the return value, just checking that the request isn't unauthorized
+        digitalTwinClient.getDigitalTwin(deviceId, BasicDigitalTwin.class);
+
+        // deliberately expire the SAS token to provoke a 401 to ensure that the digital twin client is using the shared
+        // access signature that is set here.
+        sasCredential.update(SasTokenTools.makeSasTokenExpired(serviceSasToken.toString()));
```

Comment:
wouldn't it be easier to create a token that is valid for 2-3 secs and then have it expire? or are we concerned that networking delays of > 2-3secs might occus and have the initial request time out?

Reply:
Yeah, I tried to just let the previous token expire, but I couldn't get consistent 401's from the service. I figure that this current approach that just sends a wildly out of date token would completely remove the risk for clock skew or other small timing issues.